### PR TITLE
Added stubs for generic classes

### PR DESCRIPTION
**Pull request by adaamz**

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -24,9 +24,11 @@
 		"jakub-onderka/php-parallel-lint": "^1.0",
 		"nette/application": "^2.3.0 || ^3.0.0",
 		"nette/component-model": "^2.3.0 || ^3.0.0",
+		"nette/database": "^2.3.0 || ^3.0",
 		"nette/di": "^2.3.0 || ^3.0.0",
 		"nette/forms": "^2.3.0 || ^3.0.0",
 		"nette/http": "^2.3.0 || ^3.0.0",
+		"nette/routing": "^3.0.0",
```
Comment (ondrejmirtes) [inline composer.json]: There's no need for adding these in require-dev.

File: stubs/Application/Routers/RouteList.stub [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+}
+
+namespace Nette\Application\Routers;
+
+/**
+ * @phpstan-implements \IteratorAggregate<int, \Nette\Routing\Router>
+ * @phpstan-implements \ArrayAccess<int, \Nette\Routing\Router>
+ */
+class RouteList extends \Nette\Routing\RouteList implements \Nette\Application\IRouter, \ArrayAccess, \Countable, \IteratorAggregate
```
Comment (ondrejmirtes) [inline stubs/Application/Routers/RouteList.stub]: Please add these in separate files. I don't like multiple namespace in a single file.

File: stubs/Application/UI/Component.stub [hunk 3]
```diff
@@ -0,0 +1,18 @@
+<?php
+
+namespace Nette\ComponentModel;
+
+interface IComponent
+{
+
+}
+
+namespace Nette\Application\UI;
```
Comment (ondrejmirtes) [inline stubs/Application/UI/Component.stub]: ditto

File: stubs/Database/Table/Selection.stub [hunk 4]
```diff
@@ -0,0 +1,23 @@
+<?php
+
+namespace Nette\Database;
+
+interface IRow
+{
+}
+
+namespace Nette\Database\Table;
```
Comment (ondrejmirtes) [inline stubs/Database/Table/Selection.stub]: ditto

File: stubs/Forms/Container.stub [hunk 5]
```diff
@@ -0,0 +1,21 @@
+<?php
+
+namespace Nette\ComponentModel;
+
+interface IComponent
+{
+
+}
+
+namespace Nette\Forms;
```
Comment (ondrejmirtes) [inline stubs/Forms/Container.stub]: ditto

File: stubs/Utils/ArrayList.stub [hunk 6]
```diff
@@ -0,0 +1,12 @@
+<?php
+
+namespace Nette\Utils;
+
+/**
+ * @phpstan-implements \IteratorAggregate<int, mixed>
+ * @phpstan-implements \ArrayAccess<int, mixed>
+ */
+class ArrayList implements \ArrayAccess, \Countable, \IteratorAggregate
```
Comment (ondrejmirtes) [inline stubs/Utils/ArrayList.stub]: My opinion is that this class should be generic.

Comment (ondrejmirtes) [reply stubs/Utils/ArrayList.stub]: Please also update ArrayHash, it should be generic as well.

File: composer.json (hunk 1 above)
Comment (adaamz) [reply composer.json]: Why not? We use them.

Comment (ondrejmirtes) [reply composer.json]: Where?

Comment (ondrejmirtes) [reply composer.json]: I think the build will pass if we remove all/most of them.

Comment (adaamz) [reply composer.json]: They are used in stubs.

**Comment by adaamz** [comment]
I'm not sure if correct way how to check if generics works is to check tests with phpstan, but it has some outcome :smile:

Currently I'm stuck with 
```
  Line   stubs\Application\Routers\RouteList.stub
 ------ ---------------------------------------------------------------
  9      PHPDoc tag @implements has invalid type Nette\Routing\Router.
  9      PHPDoc tag @implements has invalid type Nette\Routing\Router.
```

and 2 unresolved generics, but I think I implemented all implemented interfaces, so I have no idea what to add
```
  Line   tests\GenericStubs\GenericsPHPStanTest.php
 ------ --------------------------------------------------------------------------------------------------
  54     Method PHPStan\GenericStubs\GenericsPHPStanTest::testDatabaseTableActiveRow() return type has no value type specified in iterable type Nette\Database\Table\ActiveRow.

  59     Method PHPStan\GenericStubs\GenericsPHPStanTest::testApplicationRoutersRouteList() return type has no value type specified in iterable type Nette\Application\Routers\RouteList.
```


Any idea?

File: composer.json (hunk 1 above)
Comment (ondrejmirtes) [reply composer.json]: As I've said a few days ago in the Slack thread, classes existing outside of stubs do not have any effect on stubs themselves.

**Comment by ondrejmirtes** [comment]
I don't typically write tests for stubs, they're validated by PHPStan itself. If they prove to be wrong in real-world usage, you can send a bugfix.

**Comment by adaamz** [comment]
But this test I wrote is just usage in real world - so it should pass, or not?

**Comment by ondrejmirtes** [comment]
1) The interface/class Router isn't defined in any of the stubs.
2) ActiveRow - I think you also need to annotate the `getIterator` method.

File: composer.json (hunk 1 above)
Comment (adaamz) [reply composer.json]: I removed nette/routing and build passes, but nette/database is used in that new generics test.

**Comment by ondrejmirtes** [comment]
Thank you. I'll remove the stub test on master.

**Pull request merged**